flink-staging/flink-streaming/flink-streaming-connectors/src/main/java/org/apache/flink/streaming/connectors/kafka/api/simple/iterator/KafkaMultiplePartitionsIterator.java

```diff
@@ -71,26 +71,33 @@ public boolean hasNext() {
 	}
 
 	protected int lastCheckedPartitionIndex = -1;
+	private boolean gotNewMessage = false;
 
 	@Override
 	public MessageWithMetadata nextWithOffset() throws InterruptedException {
-		KafkaOnePartitionIterator partition;
+		KafkaSinglePartitionIterator partition;
 
 		while (true) {
 			for (int i = nextPartition(lastCheckedPartitionIndex); i < partitions.size(); i = nextPartition(i)) {
 				partition = partitions.get(i);
 
 				if (partition.fetchHasNext()) {
+					gotNewMessage = true;
 					lastCheckedPartitionIndex = i;
 					return partition.nextWithOffset();
 				}
 			}
 
-			try {
-				Thread.sleep(waitOnEmptyFetch);
-			} catch (InterruptedException e) {
-				e.printStackTrace();
+			// do not wait if a new message has been fetched
+			if (!gotNewMessage) {
+				try {
+					Thread.sleep(waitOnEmptyFetch);
+				} catch (InterruptedException e) {
+					e.printStackTrace();
```

Comment:
Never do this ;) It will print a stacktrace to stdout without a message or anything.
In that case, I would log a debug message.

I'll update it for you!